### PR TITLE
feat: use message_id instead of hash for message_filter

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/crypto.rs [hunk 1]
```diff
@@ -19,17 +19,6 @@ use xor_name::{XorName, XOR_NAME_LEN};
 /// SHA3-256 hash digest.
 pub type Digest256 = [u8; 32];
```
Comment (dirvine) [inline src/crypto.rs:21]: Can this be removed now?

File: src/message_filter.rs [hunk 2]
```diff
@@ -6,38 +6,16 @@
...
-            Self::NewMessage => true,
-            Self::KnownMessage => false,
-        }
-    }
-}
-
 // Structure to filter (throttle) incoming and outgoing messages.
 pub(crate) struct MessageFilter {
-    incoming: LruCache<MessageHash, ()>,
-    outgoing: LruCache<(MessageHash, XorName), ()>,
```
Comment (dirvine) [inline src/message_filter.rs:40]: What happened to filtering outgoing?

File: src/routing/mod.rs [hunk 3]
```diff
@@ -466,10 +466,29 @@ async fn handle_message(dispatcher: Arc<Dispatcher>, bytes: Bytes, sender: Socke
     };
     let _span_guard = span.enter();
 
-    let message_type = match WireMsg::deserialize(bytes) {
+    let wire_msg = match WireMsg::from(bytes) {
```
Comment (dirvine) [inline src/routing/mod.rs:469]: Are we deserialising the whole message here as the point is not to and just read dst an dif it's us then process it (deserialise) or forward it. 
Also msg_id to be read from msg without deserializing the whole message and added to filter / checked we have handled it.

File: src/routing/dispatcher.rs [hunk 4]
```diff
@@ -42,6 +42,10 @@ impl Dispatcher {
         }
     }
 
+    pub async fn contains_incoming(&self, msg_id: &MessageId) -> bool {
+        self.core.lock().await.contains_incoming(msg_id)
+    }
+
```
Comment (dirvine) [inline src/routing/dispatcher.rs:48]: Possibly better to have `add_to_filter() ->bool` if added then cool, if not added then drop message (it's filtered)

File: src/routing/mod.rs (hunk 3 above)
Comment (maqi) [reply src/routing/mod.rs:469]: `WireMsg::from` only restore the header part, which contains the msg_id
`WireMsg::deserialize` or `WireMsg::to_message` will deserialize the whole message

File: src/crypto.rs (hunk 1 above)
Comment (maqi) [reply src/crypto.rs:21]: we use it in DKGKey

File: src/message_filter.rs (hunk 2 above)
Comment (maqi) [reply src/message_filter.rs:40]: as each WireMsg will be given a random msg_id, there is no point to filtering outgoing (based on random msg_id).

Comment (dirvine) [reply src/message_filter.rs:40]: Why did we filter outgoing previously and if there was a valid reason are we now creating an issue? Or was it always useless?

File: src/routing/mod.rs [hunk 5]
```diff
@@ -466,10 +466,29 @@ async fn handle_message(dispatcher: Arc<Dispatcher>, bytes: Bytes, sender: Socke
...
+        Ok(wire_msg) => wire_msg,
+        Err(error) => {
+            error!("Failed to deserialize wired_message: {}", error);
```
Comment (bochaco) [inline src/routing/mod.rs:472]: I would here log something like "Failed to deserialize message header: {}"

File: src/routing/mod.rs [hunk 6]
```diff
@@ -466,10 +466,29 @@ async fn handle_message(dispatcher: Arc<Dispatcher>, bytes: Bytes, sender: Socke
...
+        );
+        return;
+    }
+
+    let message_type = match wire_msg.to_message() {
         Ok(message_type) => message_type,
         Err(error) => {
-            error!("Failed to deserialize message: {}", error);
+            error!(
+                "Failed to deserialize message({:?}): {}",
```
Comment (bochaco) [inline src/routing/mod.rs:488]: ...and may here log something like "Failed to deserialize message payload ({:?}): {}"

File: src/routing/core/messaging/handling/mod.rs [hunk 7]
```diff
@@ -62,12 +62,6 @@ impl Core {
             return Ok(commands);
         }
 
-        // Filter messages which were already handled
-        if self.msg_filter.contains_incoming(&msg) {
-            trace!("not handling message - already handled: {:?}", msg);
```
Comment (dirvine) [inline src/routing/core/messaging/handling/mod.rs:67]: Should this not still be the case?

**Review by dirvine — APPROVED** [review]

**Pull request merged**